### DISCUSSION TITLE
#[derive(FromBytes)] not checked for generic structs?

### DISCUSSION THREAD

**Discussion by peter-lyons-kehl** [google/zerocopy#2150] [General]
Hi. I must be missing something obvious, but I can't believe that the following compiles (both with `1.85.0-nightly` and `1.83.0` stable (installed with `rustup` on current x64 Manjaro Linux). Any tips, please?

`Cargo.toml` contains:
```toml
[dependencies]
zerocopy = "0.8.13"
zerocopy-derive = "0.8.13"
```

```rust
#![no_std]

use zerocopy::FromZeros;
use zerocopy_derive::*;

#[derive(FromZeros)]
pub struct WrapFromZeros<Repr: FromZeros> (Repr);

pub type WrapU8 = WrapFromZeros<u8>;
// Should fail to compile:
pub type WrapOptStr = WrapFromZeros<Option<&'static str>>;

// Intentionally no representation specified, surprised that we don't get a compile error:
enum NoRepr {}
pub type WrapNoRepr = WrapFromZeros<NoRepr>;
```

https://docs.rs/zerocopy/latest/zerocopy/derive.FromZeros.html doesn't mention anything about generics.

**Comment by peter-lyons-kehl** [comment]
I'm also perplexed that it fails even _without_ the `#[derive(FromB)]` line. I thought that the bound `Repr: FromB` would fail for `Option<&'static str>` and for `enum NoRepr {}`.

**Comment by jswrenn** [comment]
I think your confusion stems from Rust's odd treatment of type aliases: they're essentially unchecked at the definition site. For example, this produces no error at all:
```rust
struct Wrapper<T: Sized>(T);
type UnsizedWrapper = Wrapper<[u8]>;
```
...until you actually use the alias; e.g., this:
```rust
impl UnsizedWrapper {}
```
...results in an error:
```
error[E0277]: the size for values of type `[u8]` cannot be known at compilation time
 --> src/lib.rs:4:6
  |
4 | impl UnsizedWrapper {
  |      ^^^^^^^^^^^^^^ doesn't have a size known at compile-time
  |
  = help: the trait `Sized` is not implemented for `[u8]`
note: required by a bound in `Wrapper`
 --> src/lib.rs:1:19
  |
1 | struct Wrapper<T: Sized>(T);
  |                   ^^^^^ required by this bound in `Wrapper`
```

**Comment by peter-lyons-kehl** [comment]
Wow, thank you.
